File: python/client_test.py
from src import ExperimentClient
from time import sleep


def on_experiment_started(parameters):
    print("started a new experiment", parameters)


def on_episode_started(episode_number):
    print("started a new episode", episode_number)


def on_episode_finished(episode_number):
    print("ended a new episode", episode_number)


def on_experiment_finished():
    print("ended a new experiment")


def print_step(step):
    print(step)


client = ExperimentClient()
client.on_experiment_started = on_experiment_started
client.on_episode_started = on_episode_started
client.on_episode_finished = on_episode_finished
client.on_experiment_finished = on_experiment_finished
client.connect()

client.subscribe()

client2 = ExperimentClient()
client2.connect()
print("start_experiment")
response = client2.start_experiment("PREFIX", "SUFFIX", "hexagonal", "mice", "10_05", "test_subject", 10)
experiment_name = response.experiment_name
sleep(1)
for i in range(15):
    print("start_episode")
    if not client2.start_episode(experiment_name):
        print("failed to start episode")
    sleep(5)
    print("finish_episode")
    if not client2.finish_episode():
        print("failed to finish episode")
    sleep(1)
print("finish_experiment")
if not client2.finish_experiment(experiment_name):
    print("failed to finish experiment")

File: catchtests/client_test.cpp
#include <catch.h>
#include <experiment.h>

using namespace cell_world;
using namespace experiment;
using namespace std;
using namespace this_thread;

struct Client : experiment::Experiment_client {
    void on_experiment_started(const Start_experiment_response &experiment) override{
        cout << "on_experiment_started: " <<  experiment << endl;
    }
    void on_episode_started(const std::string &experiment_name) override {
        cout << "on_episode_started: " <<  experiment_name << endl;
    }
    void on_episode_finished() override{
        cout << "on_episode_finished: " << endl;
    }
    void on_experiment_finished(const std::string &experiment_name) override{
        cout << "on_experiment_finished: " <<  experiment_name << endl;
        disconnect();
    }
};

TEST_CASE("client_test") {
    auto t1 = thread ([]() {
        Client client;
        client.connect("127.0.0.1");
        client.subscribe();
        client.join();
    });
    World_info wi;
    wi.world_configuration = "hexagonal";
    wi.world_implementation = "cv";
    wi.occlusions = "10_05";
    Client client;
    client.connect("127.0.0.1");
    sleep_for(1s);
    auto experiment = client.start_experiment(wi,"test_subject",10,"prefix","suffix");
    sleep_for(1s);
    client.start_episode(experiment.experiment_name);
    sleep_for(10s);
    client.finish_episode();
    sleep_for(1s);
    client.start_episode(experiment.experiment_name);
    sleep_for(10s);
    client.finish_episode();
    sleep_for(1s);
    client.start_episode(experiment.experiment_name);
    sleep_for(10s);
    client.finish_episode();
    sleep_for(1s);
    client.start_episode(experiment.experiment_name);
    sleep_for(10s);
    client.finish_episode();
    sleep_for(1s);
    client.start_episode(experiment.experiment_name);
    sleep_for(10s);
    client.finish_episode();
    sleep_for(1s);
    client.finish_experiment(experiment.experiment_name);
    sleep_for(1s);
    client.disconnect();
    t1.join();
}